[Checkpoint] Fix autocasting

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81766

Add support for the correct autocasting in the non-reentrant checkpoint as it exists in the reentrant-version.

This was noticed by @awgu.

rohan-varma committed:
[Checkpoint] Fix autocasting

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -4457,6 +4457,35 @@ def test_checkpointing(self):
         mean_combined = torch.stack(feat_combined).mean()
         mean_combined.backward()
 
+#    @unittest.skipIf(not torch.cuda.is_available(), "Test requires CUDA")
```
This can be removed?
Should we have a second test that tests the same thing on cuda?

albanD reviewed (approved):
The fix looks good. Just a minor nit on testing

awgu commented on torch/utils/checkpoint.py:
```diff
@@ -323,7 +323,17 @@ def _checkpoint_without_reentrant(function, preserve_rng_state=True, *args):
         *args: Arguments to pass in to the given ``function``.
     """
-    had_autocast_in_fwd = torch.is_autocast_enabled()
+    # Accommodates the (remote) possibility that autocast is enabled for cpu AND gpu.
```
nit: Should we refactor this commonality into something like `_get_autocast_kwargs()` that returns these two `dict`s to avoid the duplication with reentrant AC?

awgu commented on torch/utils/checkpoint.py:
```diff
@@ -371,7 +381,12 @@ def inner_unpack(packed):
                         set_device_states(fwd_gpu_devices, fwd_gpu_states)
-                with torch.enable_grad(), torch.cuda.amp.autocast(had_autocast_in_fwd):
+
+                with (
```
nit: Stylistically, if we are doing `with ctx1, ctx2, ctx3:` already, maybe we can include the existing `saved_tensors_hooks()` context in the same `with` statement to avoid the nested `with`.

rohan-varma replied:
Does

```
with (a, b)
```

guarantee that a is entered before b?

awgu replied:
Yes. I believe so.
<img width="818" alt="Screen Shot 2022-07-20 at 1 41 30 PM" src="https://user-images.githubusercontent.com/31054793/180047496-3a68397e-908e-4f0e-97c2-ab3d96bbe29c.png">
https://docs.python.org/3/reference/compound_stmts.html#with

rohan-varma replied:
I see. Actually appears that this syntax is limited to 3.10?

awgu replied:
I think for 3.10, they let you use parentheses like:
```
with (
    A() as a,
    B() as b,
):
    SUITE
```
Already in 3.1, you can do like:
```
with A() as a, B() as b:
    SUITE
```
(For multi-line, you need to include the `\` yourself. See the second answer here: https://stackoverflow.com/questions/3024925/create-a-with-block-on-several-context-managers)

awgu commented on test/test_autograd.py:
```diff
@@ -4457,6 +4457,35 @@ def test_checkpointing(self):
+                loss = checkpoint(foo, x, y, z, use_reentrant=False)
+                loss = loss.sum()
+
+            # Without fix, would raise error in autograd about mismatched dtypes.
```
nit: might be good to specify what the fix was about since in the future, it may not be clear what "fix" refers to
```suggestion
            # Without saving and restoring the autocast dtype, would raise error in autograd about mismatched dtypes.
```

awgu reviewed (commented):
I left a few nits.

rohan-varma committed:
Update on "[Checkpoint] Fix autocasting"


Add support for the correct autocasting in the non-reentrant checkpoint as it exists in the reentrant-version.

This was noticed by awgu. 

[ghstack-poisoned]

rohan-varma committed:
Update on "[Checkpoint] Fix autocasting"


Add support for the correct autocasting in the non-reentrant checkpoint as it exists in the reentrant-version.

This was noticed by awgu. 

[ghstack-poisoned]

rohan-varma committed:
Update on "[Checkpoint] Fix autocasting"


Add support for the correct autocasting in the non-reentrant checkpoint as it exists in the reentrant-version.

This was noticed by awgu. 

[ghstack-poisoned]

rohan-varma committed:
Update on "[Checkpoint] Fix autocasting"


Add support for the correct autocasting in the non-reentrant checkpoint as it exists in the reentrant-version.

This was noticed by awgu. 

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge
